Support safe area insets to improve UX on mobile browsers.

rval opened the pull request:
Adds support for [safe area insets](https://developer.mozilla.org/en-US/docs/Web/CSS/env#values) to improve the user experience on mobile browsers.

This prevents UI elements (e.g., message text input, sidebar controls) from being obscured by the iPhone's home indicator and notch, as shown in the attached screenshots.

This tries to merge in some of the work from @mhajder's [pull request](https://github.com/thelounge/thelounge/pull/4516).

# Details

* I've checked to ensure the mentions panel is still dismissible and performs the same way it does in the main branch.
* This MR tries to follow Apple's recommendations (and the web standard) as closely as possible. That is, it uses the `safe-area-insets` as provided by the OS, without tweaks for aesthetics. The padding is consistent with the behavior of Apple's native apps on the iPhone.

# Before

Notice that the iPhone's home indicator obscures the message input box and sidebar controls, making them difficult to tap.

<img width="244" alt="Channel in Portrait (Before)" src="https://github.com/user-attachments/assets/fb1776dc-f266-46af-a2b3-755a0cf403a3">
<img width="244" alt="Sidebar in Portrait (Before)" src="https://github.com/user-attachments/assets/5f3378fb-01c9-4cd2-ba56-9ed980e5c226">
<img width="464" alt="Channel in Landscape with Keyboard (Before)" src="https://github.com/user-attachments/assets/fc2c7973-f5ca-4b69-b1ca-75d72d8d2c64">
<img width="464" alt="Channel in Landscape (Before)" src="https://github.com/user-attachments/assets/cc5a9bc9-a366-46ed-afb6-46a97218bba2">
<img width="464" alt="Sidebar in Landscape (Before)" src="https://github.com/user-attachments/assets/e6995a84-3ec9-4598-9d73-c43d0f1e899e">

# After

The iPhone's home indicator doesn't obscure the message input box or sidebar controls, and it extends into the background (at least in portrait), which is more consistent with Apple's native iPhone apps. 

<img width="244" alt="Channel Portrait (Without Chrome)" src="https://github.com/user-attachments/assets/eaf658ad-8161-4e1b-9876-7cb3c01880d1">
<img width="244" alt="Sidebar Portrait (With Chrome)" src="https://github.com/user-attachments/assets/3f96ffb8-7741-4f80-bdca-e5f3b14aa29a">
<img width="464" alt="Channel in Landscape" src="https://github.com/user-attachments/assets/ec84aef4-85d5-4624-97ff-02653f5696d3">
<img width="464" alt="Channel in Landscape with Keyboard" src="https://github.com/user-attachments/assets/a5dbfada-eec6-4114-85ac-e4f03a28d55e">

# Comparison With Native Apps

Note that the padding is consistent with native apps on the iPhone (although it looks more pronounced because the Lounge's text field doesn't have borders):

<img width="244" alt="Messages" src="https://github.com/user-attachments/assets/749c5358-48cc-499b-a34c-8bfe63b50c86">
<img width="244" alt="Channel Portrait (Without Chrome)" src="https://github.com/user-attachments/assets/eaf658ad-8161-4e1b-9876-7cb3c01880d1">

rval commented on client/index.html.tpl:
```diff
@@ -3,7 +3,7 @@
 
 	<meta charset="utf-8">
-	<meta name="viewport" content="width=device-width, user-scalable=no">
+	<meta name="viewport" content="width=device-width, user-scalable=no, viewport-fit=cover">
```
Note that `viewport-fit=cover` is required to making the safe area insets work.

brunnre8 commented:
Thanks for your effort.

There's a similar PR in https://github.com/thelounge/thelounge/pull/4516 that I can't really test as I don't own an iOS device and on Android those things work a bit differently, least on my device.

Can you compare which of the two works better for you? There was some feedback on the earlier PR, not that I like the games that are played by reducing the inset areas... that probably has to go away.
Because if devices really say what they need this would again break things.

Let people complain to apple if the margins are insane.

But there seems to have been an issue with the notification popup for the simple solution, can you evaluate which of the approaches work better for you? The older PR or yours?

rval commented on client/css/style.css:
```diff
@@ -2285,6 +2290,10 @@ part/quit messages where we don't load previews (adds a blank line otherwise) */
 
+#mentions-popup-container {
+	margin-top: env(safe-area-inset-top);
+}
```
Ensures the mentions popup container can still be dismissed.

rval commented:
@brunnre8 I wasn't aware of that pull request, thanks for letting me know. 

I think either one would be an improvement, but I tried to integrate the changes from @mhajder's PR into this one with the following changes:

* I avoided aesthetic tweaks. This PR is purely focused on fixing the usability issues. 
* I wasn't able to get safe area insets working properly in landscape without breaking portrait orientation. Since landscape was relatively usable, I focused on just getting a fix for portrait mode in place.
* I did make sure the mentions popup continues to work as it does in the main branch (at least as far as I can tell).

Let me know what you think!

brunnre8 commented:
as said, can't test it myself so I do rely on feedback from iOS folks like yourself.

All I'll do is test that you don't break it on my devices 😉 

Thanks a bunch, will have a look tomorrow

rval commented:
@brunnre8 I've updated my pull request to address some of the remaining issues and updated the description with new images. I think this PR does improve on the originals now.

Let me know if there's anything else I can do to help get this over the finish line. I took some video of the interactions but Github's rejecting the file sizes.

brunnre8 commented on client/css/style.css:
```diff
@@ -107,6 +108,13 @@ body {
+@media screen and (orientation: portrait) {
+	/* Allow the main chat interface to merge into the background. */
+	body {
+		background: white;
```
please remove that again.
I don't see why you need to set the color there for a padding issue.

People use dark themes you know and having some flash white is seldom a good idea :)

rval replied:
This was to get the effect mentioned in my other comment, but definitely not necessary. Removed as requested.

brunnre8 commented on client/css/style.css:
```diff
@@ -2211,6 +2219,18 @@ part/quit messages where we don't load previews (adds a blank line otherwise) */
+	}
+
+	/* When the form becomes focused, remove the safe area insets padding. A bit of a hack, but seems to work. */
+	#form:focus-within {
```
why exactly is this needed now? Your first version didn't have it?

rval replied:
Without this, the keyboard keeps the safe area padding when it's focused, which looks weird and reduces the limited amount of vertical space. 

Here's what it looks like with this CSS (notice the extra padding in the input box):

<img width="422" alt="Padded Form" src="https://github.com/user-attachments/assets/cda5317b-c4f7-47d8-b8a3-9efb6b0b98b1">

And here it is with the CSS (notice the padding is not there, and there's more room for the channel content):

<img width="422" alt="Unpadded Form" src="https://github.com/user-attachments/assets/e1c388c1-4657-4d5f-bd11-b307c8b304be">

I'm not sure why it wasn't present in my earlier version. I tried a bunch of different ways to address this but this was the best one I landed on.

rval replied:
@brunnre8 if this is a blocker (or we just want to kick this particular decision down the road) I could make one of the following changes:

1. I can remove the `#form:focus-within` hack, and we can just let the padding exist as shown in the screenshot above.
2. I can descope this pull request to support the safe area insets _only_ for portrait orientation, and leave landscape orientation as is (it can always be tackled later since portrait mode is the bigger issue, I think).

Let me know if you have a preference.

brunnre8 commented on client/themes/default.css:
```diff
@@ -15,6 +15,12 @@
 
+@media screen and (orientation: portrait) {
+	.window {
+		box-shadow: 0 0 0;
```
why are you messing with the shadows here?

rval replied:
It was an attempt to make it look a bit more at home on iOS. Here it is with and without these rules, respectively:

<img width="244" alt="Channel in Portrait" src="https://github.com/user-attachments/assets/b31a1215-e0d7-4f40-89eb-ca7a26b0a4d2">
<img width="244" alt="Channel Portrait (Without Chrome)" src="https://github.com/user-attachments/assets/eaf658ad-8161-4e1b-9876-7cb3c01880d1">

But probably better left to CSS overrides or a separate theme maybe, so I left it out. I'll update the screenshots in the merge request to reflect the new appearance.

JipFr replied:
I definitely think that the background should be the same color as the text input box, perhaps there is a different way to enforce that?

ivanjx commented:
what is still blocking the merge? even though it might not be perfect this is still way better than the current ux

FlyGoat commented:
+1, this improved iOS PWA experience a lot.

macmanx2 commented:
Especially after iOS 26, this would _dramatically_ improve The Lounge as a PWA app. Right now, the iPhone's home indicator takes up the bottom 1/4 of the text box, meaning that you'll need to be very precise when tapping in to type, or you might launch Siri.

macmanx2 commented:
While all of the above is the right way to fix this, until/if it's all merged (and until/if your admin updates), I "fixed" it by adding this to my custom stylesheet (adjust as desired).

`#form {
    padding-bottom: 16px;
}`

benyafai commented:
> While all of the above is the right way to fix this, until/if it's all merged (and until/if your admin updates), I "fixed" it by adding this to my custom stylesheet (adjust as desired).
> 
> 
> 
> `#form {
> 
>     padding-bottom: 16px;
> 
> }`



I've had this in my instance for years; 

```
#form {
  margin-bottom: 30px;
}
textarea#user-specified-css-input {
  height: 600px;
}
```

balgerion commented:
Any chance to merge this PR ? we are now in iOS 26 and rly need this to have working PWA
